yarn-project/merkle-tree/src/snapshots/snapshot_builder_test_suite.ts

```diff
@@ -0,0 +1,138 @@
+import { TreeBase } from '../tree_base.js';
+import { SnapshotBuilder } from './snapshot_builder.js';
+
+/** Creates a test suit for snapshots */
+export function describeSnapshotBuilderTestSuite<T extends TreeBase, S extends SnapshotBuilder>(
+  getTree: () => T,
+  getSnapshotBuilder: () => S,
+  modifyTree: (tree: T) => Promise<void>,
+) {
+  describe('SnapshotBuilder', () => {
+    let tree: T;
+    let snapshotBuilder: S;
+    let leaves: bigint[];
+
+    beforeEach(() => {
+      tree = getTree();
+      snapshotBuilder = getSnapshotBuilder();
+
+      leaves = Array.from({ length: 4 }).map(() => BigInt(Math.floor(Math.random() * 2 ** tree.getDepth())));
+    });
+
+    describe('snapshot', () => {
+      it('takes snapshots', async () => {
+        await modifyTree(tree);
+        await tree.commit();
+        await expect(snapshotBuilder.snapshot(1)).resolves.toBeDefined();
+      });
+
+      it('refuses to take the same snapshot twice', async () => {
+        await modifyTree(tree);
+        await tree.commit();
+
+        const version = 1;
+        await snapshotBuilder.snapshot(version);
+        await expect(snapshotBuilder.snapshot(version)).rejects.toThrow();
```

Comment:
Not clear to me if we want it to throw or just to return the one it already made?

Reply:
Oh good shout!